### PR TITLE
AnkiDroid 2.15: Downgrade V16 schema database

### LINKED ISSUE #8529: Implement Downgrade & Quit [in progress in Anki-Backend repo]
If we can "Downgrade & Quit" in 2.15, this will save us a lot of hassle with the database schema upgrade in 2.16

This will allow seamless downgrades from 2.16 to 2.15

Open to discussion for how we present this option to users

### PR DESCRIPTION
**Pull request by david-allison**
## Purpose / Description

* A user on AnkiDroid 2.16 would be on a V16 schema, opening this collection in v2.15 would be on v15, and this fails as the collection tables have been wiped

We want to allow this downgrade to allow alpha/beta users to straddle the releases. General users who downgrade will benefit as well

## Fixes
Fixes #8529 

## Approach
* The anki backend has been upgraded to expose a "downgrade collection" endpoint
* Check for v16
* If it exists: back up
* Then if the backup succeeds, call this endpoint
* Then continue as normal if it succeeds

## How Has This Been Tested?

Unit tested, and tested on my device (Android 11) with a dummy collection

## Learning (optional, can help others)
https://github.com/david-allison-1/anki/blob/6b2410e0ba22422970f861697a767150ec269637/rslib/src/storage/sqlite.rs#L200-L206

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by david-allison-1** [comment]
I feel that this is likely too large to review. 

Given the complexity of the change, please review carefully and let me know how you'd like things split up

File: AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java [hunk 1]
```diff
@@ -340,6 +347,96 @@ public void onResumeLoadCollectionSuccessWithAccessibleCollection() {
...
+            assertThat("An error dialog should be displayed", deckPicker.mDatabaseErrorDialog, is(DatabaseErrorDialog.DIALOG_LOAD_FAILED));
+        } finally {
+            InitialActivityTest.setupForDefault();
+        }
+    }
+
+
+    private void useCollection(@SuppressWarnings("SameParameterValue") CollectionType collectionType) {
+        // load asset into temp
+        String path = ResourceLoader.getTempCollection(getTargetContext(), collectionType.getAssetFile());
```
Comment (mikehardy) [inline AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java:413]: nifty!

File: AnkiDroid/src/main/java/com/ichi2/libanki/Consts.java [hunk 2]
```diff
@@ -122,6 +122,8 @@
 
     // deck schema & syncing vars
     public static final int SCHEMA_VERSION = 11;
+    /** The database schema version that we can downgrade from */
+    public static final int SCHEMA_VERSION_DOWNGRADE_SUPPORTED = 16;
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/libanki/Consts.java:126]: this seems silly but I was confused by this based on the mental expectation I had from the name of the constant. I was looking for a boolean or something (supported? yes no) and had trouble parsing the conditional above where you decide if you it is supported or not. Maybe reverse?

`DOWNGRADE_SUPPORTED_SCHEMA_VERSION` ? Now I'm looking for a number mentally 

Not sure what you have in mind for future but maybe even

`DOWNGRADE_SUPPORTED_SCHEMA_VERSION_MAX` or similar so now I'm looking for a number and I have the idea it's a range I might be outside of

File: AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java [hunk 3]
```diff
@@ -389,6 +389,24 @@ public void corruptVersion16CollectionShowsDatabaseError() {
...
+        try {
+            setupColV16();
+
+            InitialActivityTest.setupForValid(getTargetContext());
+
+            DeckPickerNoSpaceForBackup deckPicker = super.startActivityNormallyOpenCollectionWithIntent(clazz, new Intent());
+            waitForAsyncTasksToComplete();
+
+            assertThat("Collection should not be open", !CollectionHelper.getInstance().colIsOpen());
+            assertThat("An downgrade failed dialog should be shown", deckPicker.mDisplayedDowngradeFailed, is(true));
```
Comment (mikehardy) [inline AnkiDroid/src/test/java/com/ichi2/anki/DeckPickerTest.java:404]: nit -> `An` -> `A`

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java [hunk 4]
```diff
@@ -1297,6 +1313,12 @@ private void showStartupScreensAndDialogs(SharedPreferences preferences, int ski
         }
     }
 
+    public void displayDowngradeFailedNoSpace() {
+        Timber.w("Not enough space to downgrade");
+        showDialogFragment(DeckPickerNoSpaceToDowngradeDialog.newInstance());
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java:1318]: This isn't actionable for the user, any time I am on a space-constrained device, it really really helps me to know how much space is needed, how much is available and how much remains to be freed for it to succeed. Then I have a goal for freeing space and can free the requirement amount

Is it possible to plumb through the space required for the downgrade (or the file being downgraded)? I *assume* it's the size of the collection being downgraded, so with that info we can construct an actionable message

**Review by mikehardy — COMMENTED** [review]
I thought review was fine, it was chunked up about as much as I think I can. I noted like one extraction that you did in situ but it was no bother

The only thing I can think of is making the out of space dialog actionable as noted

**Comment by mikehardy** [comment]
exciting to get this in - we're going to have run through the blocker list pretty shortly - thanks!

**Comment by david-allison-1** [comment]
image

**Review by mikehardy — APPROVED** [review]
I think it's great. It is big, and it's all over your area, I approve merge at any time but I know when I author big PRs sometimes I rescan them a few more times before merge and want to polish something. Merge at will

**Comment by mikehardy** [comment]
Merging after confirm it was :+1: from David

**Comment by mikehardy** [comment]
Hi there!

Apologies for the delay, and if you're already submitted something to OpenCollective for May 2021 PRs merged, you may feel free to ignore this.

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for May 2021 submissions

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

(I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged in May)

Thanks!

**Pull request merged**